src/background-view.cpp

```diff
@@ -0,0 +1,355 @@
+
+    int set_cloexec_or_close(int fd)
+    {
+        if (os_fd_set_cloexec(fd) != 0)
+        {
+            close(fd);
+            return -1;
+        }
+        return fd;
+    }
+
+    int os_socketpair_cloexec(int domain, int type, int protocol, int *sv)
+    {
+        int ret;
+
+#ifdef SOCK_CLOEXEC
+        ret = socketpair(domain, type | SOCK_CLOEXEC, protocol, sv);
+        if (ret == 0 || errno != EINVAL)
+        {
+            return ret;
+        }
+#endif
+
+        ret = socketpair(domain, type, protocol, sv);
+        if (ret < 0)
+        {
+            return ret;
+        }
+
+        sv[0] = set_cloexec_or_close(sv[0]);
+        sv[1] = set_cloexec_or_close(sv[1]);
+
+        if (sv[0] != -1 && sv[1] != -1)
+        {
+            return 0;
+        }
+
+        close(sv[0]);
+        close(sv[1]);
+        return -1;
+    }
+
+    static int sigchld_handler(int signal_number, void *data)
+    {
+        std::map<wf::output_t *, struct process> *procs = (std::map<wf::output_t *, struct process> *) data;
+        pid_t pid;
+        int status;
+
+        while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
+            for (auto& p : *procs)
+            {
+                if (p.second.pid == pid)
+                {
+                    p.second.view = nullptr;
+                    break;
+                }
+            }
+        }
+
+        if (pid < 0 && errno != ECHILD)
+        {
+            LOGE("waitpid error: ", strerror(errno));
+        }
+
+        return 1;
+    }
+
+    void child_client_exec(int sockfd, const char *command)
+    {
+        int clientfd;
+        char s[32];
+        sigset_t allsigs;
+
+        /* do not give our signal mask to the new process */
+        sigfillset(&allsigs);
+        sigprocmask(SIG_UNBLOCK, &allsigs, NULL);
+
+        /* Launch clients as the user. Do not launch clients with wrong euid. */
+        if (seteuid(getuid()) == -1)
+        {
+            LOGE("seteuid failed");
+            return;
+        }
+
+        /* SOCK_CLOEXEC closes both ends, so we dup the fd to get a
+         * non-CLOEXEC fd to pass through exec. */
+        clientfd = dup(sockfd);
+        if (clientfd == -1)
+        {
+            LOGE("dup failed: ", strerror(errno));
+            return;
+        }
+
+        snprintf(s, sizeof s, "%d", clientfd);
+        setenv("WAYLAND_SOCKET", s, 1);
+        setenv("WAYLAND_DISPLAY", wf::get_core().wayland_display.c_str(), 1);
+
+        setenv("_JAVA_AWT_WM_NONREPARENTING", "1", 1);
+
+#if WLR_HAS_XWAYLAND
+        if (wf::get_core().get_xwayland_display() >= 0)
+        {
+            auto xdisp = ":" + std::to_string(wf::get_core().get_xwayland_display());
+            setenv("DISPLAY", xdisp.c_str(), 1);
+        }
+#endif
+
+        if (execlp("bash", "bash", "-c", command, NULL) < 0)
+        {
+            LOGE("executing '", command ,"' failed: ", strerror(errno));
+        }
+    }
+
+    struct wl_client *client_launch(const char *command)
```

Comment:
I wonder, what will happen if we put this in core? Like, there is a `run()` function, maybe we can make it return the PID of the opened program.

Reply:
Is there a way to get the PID of a native client? Currently it relies on matching the wl_client for native views and PID for xwayland clients.

Reply:
As discussed on IRC, we can use wl_client_get_credentials() to get the pid of native clients. The patch for wayfire is here http://ix.io/2e0S